src/modules/lost/functions.c

```diff
@@ -258,17 +282,95 @@ int lost_held_function(struct sip_msg *_m, char *_con, char *_pidf, char *_url,
 	LM_DBG("held location request: [%s]\n", que.s);
 
 	/* send locationRequest to location server - HTTP POST */
-	curlres = httpapi.http_connect(_m, &con, NULL, &res, mtheld, &que);
+	if(con.s != NULL && con.len > 0) {
+
+		LM_DBG("using connection [%.*s]\n", con.len, con.s);
+
+		/* send via connection */
+		curl = httpapi.http_connect(_m, &con, NULL, &res, mtheld, &que);
+	} else {
+		/* we have no connection ... do a NAPTR lookup */
+		if(lost_parse_host(did.s, &host, &flag) > 0) {
+
+			LM_DBG("no conn. trying NATPR lookup [%.*s]\n", host.len, host.s);
+
+			/* remove '[' and ']' from string (IPv6) */
+			if(flag == AF_INET6) {
+				host.s++;
+				host.len = host.len - 2;
+			}
+			/* is it a name or ip ... check nameinfo (reverse lookup) */
+			len = 0;
+			ipstr = lost_copy_string(host, &len);
+			if(len > 0) {
+				name.s = &(istr[0]);
+				name.len = NI_MAXHOST;
+				if(lost_get_nameinfo(ipstr, &name, flag) > 0) {
+
+					LM_DBG("ip [%s] to name [%.*s]\n", ipstr, name.len, name.s);
+
+					/* change ip string to name */
+					host.s = name.s;
+					host.len = name.len;
+				} else {
+
+					/* keep string */
+					LM_DBG("no nameinfo for [%s]\n", ipstr);
+				}
+				pkg_free(ipstr); /* clean up */
+			}
+			url.s = &(ustr[0]);
+			url.len = MAX_URI_SIZE;
+			if((naptr = lost_naptr_lookup(host, &sheld, &url)) == 0) {
+				LM_ERR("NAPTR failed on [%.*s]\n", host.len, host.s);
+				lost_free_string(&que); /* clean up */
+				lost_free_string(&idhdr);
+				goto err;
+			}
+		} else {
+			LM_ERR("failed to get location service for [%.*s]\n", did.len,
+					did.s);
+			lost_free_string(&que); /* clean up */
+			lost_free_string(&idhdr);
+			goto err;
+		}
+
+		LM_DBG("NATPR lookup returned [%.*s]\n", url.len, url.s);
+
+		/* curl doesn't like str */
+		len = 0;
+		lisurl = lost_copy_string(url, &len);
```

Comment:
Its necessary to make a copy here, can not the url.s be used?

Reply:
curl requires a zero terminated string as url, which is not added by the http_client API query function `http_client_query_c`